tasks/licenses.js

```diff
@@ -10,75 +8,85 @@ export default function licenses(grunt) {
+      npmLicense.init(options, (result, error) => {
         cb(undefined, result);
       });
     });
 
-    // Only check production NPM dependencies, not dev
-    await fromNode(cb => npm.load({production: true}, cb));
-
     const npmList = await fromNode(cb => {
       npm.commands.list([], true, (a, b, npmList) => cb(undefined, npmList));
     });
 
-    // Recurse npm --production --json ls output, create array of dependency
-    // objects, each with a name prop formatted as 'package@version' and a
-    // path prop leading back to the root dependency.
-    function getDependencies(dependencies, path, list = []) {
-      Object.keys(dependencies).forEach(name => {
-        const dependency = dependencies[name];
-        const newPath = path ? `${path} -> ${dependency.from}` : dependency.from;
-        list.push({
-          path: newPath,
-          name: name + '@' + dependency.version
-        });
+    function flattenDependencyTree(deps, path = [], list = {}) {
+      Object.keys(deps).forEach(name => {
+        let dependency = deps[name];
+        let displayName = `${name}@${dependency.version}`;
+
+        list[displayName] = {
+          name: name,
+          version: dependency.version,
+          path: path
+        };
 
         if (dependency.dependencies) {
-          getDependencies(dependency.dependencies, newPath, list);
+          flattenDependencyTree(dependency.dependencies, path.concat(displayName), list);
         }
       });
+
       return list;
-    };
+    }
 
-    const productionDependencies = {};
-    getDependencies(npmList.dependencies).forEach(dependency => {
-      productionDependencies[dependency.name] =
-        Object.assign({}, allDependencies[dependency.name], dependency);
-    });
+    const allDependencies = flattenDependencyTree(npmList.dependencies);
+
+    /**
+     * Licenses for a package by name with overrides
+     *
+     * @param {String} name
+     * @return {Array}
+     */
+
+    function licensesForPackage(name) {
+      let licenses = packages[name].licenses;
+
+      if (config.overrides.hasOwnProperty(name)) {
+        licenses = config.overrides[name];
+      }
+
+      return typeof licenses === 'string' ? [licenses] : licenses;
+    }
+
+    /**
+     * Determine if a package has a valid license
+     *
+     * @param {String} name
+     * @return {Boolean}
+     */
+
+    function isInvalidLicense(name) {
+      let licenses = licensesForPackage(name);
+      return _.intersection(licenses, config.licenses).length !== licenses.length;
+    }
+
+    /**
+     * Generator for iterating over packages with invalid licenses
+     */
 
-    const licenseStats = _.map(productionDependencies, processPackage);
-    const invalidLicenses = licenseStats.filter(pkg => !pkg.valid);
+    let invalidLicenses = [];
 
-    if (!grunt.option('only-invalid')) {
-      grunt.log.debug(JSON.stringify(licenseStats, null, 2));
+    for (let name of Object.keys(packages)) {
```

Comment:
The functional approach means you're dealing exclusively with immutable data rather than mutating an accumulator.

Imperative approaches are fine if they are more expressive than the alternative, but that's rarely the case.  Imperative code by design is about focussing on the implementation details of the solution rather than expressing the intention of the code, which is why in this case any time you wanted to reference the current package you needed to manually wire it up by name rather than simply having a reference to it.

And don't get me wrong: my example isn't without it's own smells, like closing over `allDependencies` so we could do lookups.